Strm model evaluation

RAtshan opened the pull request.

Commit:
created a column for the difference in temperature measurements. then found the mae by summing the error and dividing by n of row for the filtered data.

Commit:
Creating metrics and facing an error with mare_metric.

Commit:
Lines (25 - 27): found the difference between
 the predicted and observed temperature (error).

Lines (30 - 32): Calculated the mae error for both uncalibrated and hybrid models.

Lines (35 - 36) : Calculated the rmse error for both uncalibrated and hybrid models.

lines (39 - 43): Calculated the mare error for both uncalibrated and hybrid models.

lines (45 - 54) Calculated the mare error for both uncalibrated and hybrid models. In this case the observed temperature values that equal to 0 were turned to a very small value.

Lines (56 - 63) Calculated the mare error for both uncalibrated and hybrid models. In this case the we removed all of the `-inf, inf`  values.

Lines (65 - 67): we combined the calculated metrics into one table for comparison.

Lines (70 - 87): plotting

Commit:
Creating function to calculate mae, rmse, and mare.

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -0,0 +1,64 @@
+  rgnc_dat_filter %>%
+    mutate( diff_observed_predicted = observe_data - predict_data ) %>%
+    group_by(seg_id_nat) %>%
+    summarize( rmse = round( sqrt( mean( diff_observed_predicted ^2, na.rm = TRUE ) ) , 2)  )
```
Some general style feedback that doesn't impact _what_ the code is doing, but does make the code more readable and standardized across our group. Don't include spaces after a `(` or before `)` within a function. Generally if you have a nested function where the line ends in `))`, there is no space between.  I would write the summarize line as:

`summarize(rmse = round(sqrt(mean(diff_observed_predicted ^ 2, na.rm = TRUE )) , 2))`

I think we largely follow [Hadley Wickam's style guide](http://adv-r.had.co.nz/Style.html).

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -0,0 +1,64 @@
+colnames(mae)[3] <- "MAE_Hybrid_Model"
+# Absolute Residual Error: Root Mean Square Error (RMSE) square the difference in predicted and observed temperature.  sum the squared values, then divided by n of row. Finally,took the square root
+cal_rmse <- function(observe_data, predict_data){
+  rgnc_dat_filter %>%
```
You don't have `rgnc_dat_filter` as an argument in your function, so the function won't generalize to other scripts/datasets. Instead, I think we should do the data manipulation outside of the function, and make the call to the function _within_ a summarize call. So something like this: 

```
cal_rmse <- function(observe_data, predict_data) {
  rmse <- sqrt(mean((observe_data - predict_data) ^ 2))
  return(rmse)
}
```
and then you would use the function like:
```
seg_rmse <- rgnc_dat_filter %>%
   group_by(seg_id_nat) %>%
   summarize(rmse = cal_rmse(temp_c, rgcn2_full_temp_c))
```

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -0,0 +1,64 @@
+}
+rmse_process_model <- cal_rmse(rgnc_dat_filter$temp_c, rgnc_dat_filter$sntemp_temp_c)
+rmse_hybrid_model <- cal_rmse(rgnc_dat_filter$temp_c, rgnc_dat_filter$rgcn2_full_temp_c)
+rmse <- left_join(rmse_process_model, rmse_hybrid_model, by = "seg_id_nat")
```
Currently, `cal_rmse` is only producing a single value, I think because the dataframe is not an argument, the function does not know how to group because it doesn't have access to those data. I think my changes above will allow you to `group_by` and then this join will work.

Commit:
Implemented Sam's suggestion to the functions. `cal_mare` is producing different values that the ones we produce in the original script.

Commit:
In the mare metric we're turning the zero values to small value (0.1). This will replace removing the `inf` values from the metric calculation.

Commit:
Lines (47 - 49): In the mare metric we're turning the zero values to small value (0.1) instead of (0.001). The change is more suitable for our data values and it provides a better metric values.

Lines (72 - 81): we used the mare metric that replaces the zero values with 0.1 instead of the metric that removes the `inf` values.

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -49,25 +49,21 @@ rmse_hybrid_model <- rgnc_dat_filter %>%
 mare_process_model <- rgnc_dat_filter %>%
   group_by(seg_id_nat) %>%
-  filter_all(all_vars(!is.infinite(.) ) ) %>%
+  mutate(temp_c = ifelse( temp_c %in% 0, 0.1, temp_c) ) %>%
```
Can you take a stab at modifying the code to match the style guide I linked to? I would modify this line to: 

`mutate(temp_c = ifelse(temp_c %in% 0, 0.1, temp_c)) %>%`

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -49,25 +49,21 @@ rmse_hybrid_model <- rgnc_dat_filter %>%
   group_by(seg_id_nat) %>%
-  filter_all(all_vars(!is.infinite(.) ) ) %>%
+  mutate(temp_c = ifelse( temp_c %in% 0, 0.1, temp_c) ) %>%
+  #filter_all(all_vars(!is.infinite(.) ) ) %>%
```
I think you can reduce redundancy by doing the `group_by` and `mutate` before you call these functions, so you're operating on the same data.

limnoliver replied:
Also, below, you could avoid the renaming of columns if in your summarize call, you named the metric what you ultimately want it to be called, e.g., 

`summarize(MARE_Hybrid_Model = round(cal_mare(temp_c, rgcn2_full_temp_c), 2))`

Commit:
Following the coding standards and removing spaces.
Naming the metric in the `summarize` function.

Grouping the data by `seg_id_nat` to avoid repetition.

Commit:
Following the coding standards and removing spaces.

Lines (57 - 62): Applying the filtering to the column of interest.

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -0,0 +1,62 @@
+}
+
+mae_process_model <- rgnc_by_seg %>% 
+  summarize(MAE_Process_Model = round(cal_mae(temp_c, sntemp_temp_c), 2))
```
I think you should put the round in the function, so you don't have to repeat it. You could even have number of digits as an input to the function, with a default of 2, but with the flexibility to change.

limnoliver replied:
Additionally, you could simplify some of this code by doing two summarizes in the pipeline, and calling the new data `mae` or something like: 
```
mae <- rgnc_by_seg %>% 
  summarize(MAE_Process_Model = round(cal_mae(temp_c, sntemp_temp_c), 2),
                     MAE_Hybrid_Model = round(cal_mae(temp_c, rgcn2_full_temp), 2))
```

Commit:
Implementing some changes to the function to avoid repetition.

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -1,62 +1,46 @@
+            MAE_Hybrid_Model = cal_mae(temp_c, rgcn2_full_temp_c))
 
 # Absolute Residual Error: Root Mean Square Error (RMSE) square the difference in predicted and observed temperature.  sum the squared values, then divided by n of row. Finally,took the square root
 cal_rmse <- function(observe_data, predict_data){
```
I think you should include n_digits as an argument to the function. The default could be set to 2 so you don't have to use it often. The function line would be:

`cal_rmse <- function(observe_data, predict_data, n_digits = 2) {`

and in the function would look like:

`rmse <- round(sqrt(mean((observe_data - predict_data) ^2, na.rm = TRUE )), digits = n_digits)`

limnoliver commented:
Looking really good, Rasha! Per our Teams discussion, the last things to add to this PR are:

1) Use some metric to compare maximum temperature in obs and modeled. Be sure you're only comparing to `seg_id_nat`s that have been completely observed over the summer period (see code snippet I sent).
2) Also calculate the timing (day of year) of that maximum temperature, and compare to modeled timing.
3) Calculate the Nash-Sutcliffe Efficiency - I sent a snippet of a calculation I've made before.

Commit:
Setting the number of rounded digits to 2. and Giving the user the option to select the number of digits when rounding the error.

Commit:
Grouped data by segment  and year.

Commit:
Grouped data by segment and year.

Lines (50 - 74):  found the maximum temperatures. Calculated the error between observe maximum and predicted maximum for each segment. Finally, found the median for error.

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -44,3 +48,27 @@ mare <- rgnc_by_seg %>%
+            max_timing_hyb = lubridate::yday(date[which.max(rgcn2_full_temp_c)]),
+            max_temp_hyb_dat = date[which.max(rgcn2_full_temp_c)],
+            error_obs_proc = abs(max_temp_c - max_temp_proc),
+            error_obs_hyb = abs(max_temp_c - max_temp_hyb),
```
what about timing error? So `max_timing_tempc` - `max_timing_proc` and for the hybrid model?

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -44,3 +48,27 @@ mare <- rgnc_by_seg %>%
+  mutate(dev_mean_obs_proc = error_obs_proc - mean(error_obs_proc),
+         dev_mean_obs_hyb = error_obs_hyb - mean(error_obs_hyb))
+
+median_metric <-data.frame(median(max_temp_timing$error_obs_proc),
```
Here you can name the columns within the `data.frame` function, so something like:

`data.frame(process_median_max_temp_error = ...`

Commit:
Finding timing error.

RAtshan reviewed (commented):
An error to inform me that filter isn't applied. Also, I wasn't sure where to apply the date_range in the function.

Commit:
Line (89-91): Added the maximums for the hybrid model.

Line (95-...) : Calculated the nash metric then changed these calculation into a function.

RAtshan reviewed (commented):
I noticed that the (nse values > 1). According to the metrics paper, "nash measurements are between -inf and one).  
Also, `nse = -inf` in line 71 (seg_id 1477 in the year 2010).

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -61,14 +61,35 @@ max_temp_timing <- rgnc_by_seg_date %>%
+  mutate(dev_mean_obs_proc = error_temp_obs_proc - mean(error_temp_obs_proc),
+         dev_mean_obs_hyb = error_temp_obs_hyb - mean(error_temp_obs_hyb))
+
+cal_max_temp_timing <- function(observe_data, predict_data, date) {
```
Because the max and max timing have a lot of redundant calculations, I suggest combining these two functions, where the function returns both the magnitude and timing metrics.

limnoliver replied:
Instead of `date` as an argument, I would call it `doy_test` and set it equal to 170:245. Then, within the function, you would have something like:

`summer_complete = all(doy_test %in% lubridate::yday(date)) `

limnoliver replied:
Also, I'm not sure what lines of code you're seeing the error. Can you insert a comment on the line where it's happening, and paste the error output?

RAtshan replied:
The code:
 `max_temp_timing_fun <- rgnc_by_seg_date %>% 
  summarize(max_obs_procc = cal_max_temp_timing(temp_c, sntemp_temp_c),
            max_obs_hyb = cal_max_temp_timing(temp_c, rgcn2_full_temp_c))`

Below is the screenshot of the error when I call the function:
![image](https://user-images.githubusercontent.com/65974907/87075572-2d028480-c1e6-11ea-91e3-ffdb8dbbfc43.png)

Commit:
Trying to create a exceedence metrics that return a logical impression.

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -101,3 +101,43 @@ nse =rgnc_by_seg_date %>%
+            hyb_test =  cal_exceedance(temp_c, rgcn2_full_temp_c))
+
+
+
```
I think you needed to return a single value per reach when using the grouped data, and the `cal_exceedance` function was not doing that. 

Here's a starter function that I think will get you going. I think we need to expand it to calculate `true_neg` and `true_pos`, too, because the high number of values below the threshold will make the proportions really small. 

```
calc_exceedance <- function(observe_data, predict_data, metric, temp_threshold = 25.5){
  observe_exceeds <- observe_data > temp_threshold
  predict_exceeds <- predict_data > temp_threshold

  if (metric == 'prop_correct') {
    return(round(sum(observe_exceeds == predict_exceeds)/n(), 2))
  } else if (metric == 'prop_false_neg') {
    return(round(sum(observe_exceeds == TRUE & predict_exceeds == FALSE)/n(), 2))
  } else if (metric == 'prop_false_pos') {
    return(round(sum(observe_exceeds == FALSE & predict_exceeds == TRUE)/n(), 2))
  }
}

exceeds <- rgnc_dat_filter %>%
  summarize(n = n(),
            n_exceeds = sum(temp_c > 25.5),
            process_exceeds_true = calc_exceedance(temp_c, sntemp_temp_c, 'prop_correct'),
            hybrid_exceeds_true = calc_exceedance(temp_c, rgcn2_full_temp_c, 'prop_correct'),
            process_exceeds_false_neg = calc_exceedance(temp_c, sntemp_temp_c, 'prop_false_neg'),
            hybrid_exceeds_false_neg = calc_exceedance(temp_c, rgcn2_full_temp_c, 'prop_false_neg'),
            process_exceeds_false_pos = calc_exceedance(temp_c, sntemp_temp_c, 'prop_false_pos'),
            hybrid_exceeds_false_pos = calc_exceedance(temp_c, rgcn2_full_temp_c, 'prop_false_pos'))
```

Commit:
Line (53 - 71): fixed the max function and produced the desired max temperature and timing metrics.

Line (81 - end): worked to fix the exceedence function but the first one `cal_exceedance_NOT` gives an error when running it. and the second one `cal_exceedance` runs without errors it doesn’t provide the desired results.

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -50,74 +50,51 @@ compare_metric <- plyr:: join_all(list(mae, rmse, mare),  by = 'seg_id_nat', typ
-cal_nash <- function(observe_data, predict_data, n_digits = 2){
-  nash = round(1 - (sum(observe_data - predict_data) ^ 2) / (sum(observe_data - mean(observe_data) ^2)), digits = n_digits)  
+cal_nash <- function(observe_col, predict_col, n_digits = 2){
+  nash = round(1 - (sum(observe_col - predict_col) ^ 2) / (sum(observe_col - mean(observe_col) ^2)), digits = n_digits)  
```
Any ideas about what's going on with the >1 values? I would suspect an order of operations bug on this line. Maybe break the line down operation by operation to make sure it's doing what it's supposed to.

limnoliver replied:
```
cal_nash <- function(observe_col, predict_col, n_digits = 2){
  nash = round(1 - ((sum((observe_col - predict_col) ^ 2)) / (sum((observe_col - mean(observe_col)) ^2))), digits = n_digits)  
}
```

Commit:
Line 53: removed unnecessary command.

Line (76 - 78): fixed the nash metric values by making sure the order of operation is correct.

Line (79 - end): fixed the exceedance metric function to produce the desire output.

limnoliver commented on code/starter_script - Copy.R:
```diff
@@ -69,40 +67,50 @@ max_process_metrics = cal_max_temp_timing(data_in = rgnc_by_seg_date,
+  observe_exceeds <- observe_col > temp_threshold
+  predict_exceeds <- predict_col > temp_threshold
+  
+  if (metric == 'prop_correct') {
```
Now you can get rid of the `prop_correct` calculation since you've covered that with `prop_true_pos` and `prop_true_neg`

Commit:
Created a separate script for the metrics functions. Documented what is the function calculating, how the calculation is made, and the explained function’s parameters.

Commit:
Removed the functions to the function script. Some functions were renamed in the function script and I had to change the function call to the new renamed function.

limnoliver commented:
The functions and documentation are looking great, @RAtshan! See my comments about cleaning up the documentation, but otherwise, we are close!

Commit:
Clarified small details and added a `@return` to the function's documentation.

Commit:
Changed the maximum function name.
